### ISSUE TITLE
Understanding code snippet on mask decoding

### ISSUE THREAD

**Issue by buttercutter** [A4Bio/ProDesign#7]
I have some problem understanding the [following code snippet](https://github.com/A4Bio/PiFold/blob/f111d66ed4c184b9614ed670da26469dcc35e588/methods/prodesign_model.py#L130-L141) around `permutation_matrix_reverse` and `mask_bw` :

here is the text analysis performed by chatGPT:

> The permutation_matrix_reverse is a binary matrix that encodes the permutation order for decoding. Each row of the matrix corresponds to one element in the output sequence. The value of the element in the i-th row and j-th column of the matrix is 1 if the i-th element in the output sequence is the j-th element in the input sequence, and 0 otherwise.
>
> The outer product of permutation_matrix_reverse with itself computes all possible pairs of elements in the output sequence, and marks each pair as "forward" or "backward". The resulting matrix order_mask_backward is a binary tensor where each element (i, j, k) corresponds to the mask for decoding the k-th element in the input sequence given the i-th and j-th elements in the output sequence. Specifically, order_mask_backward[i, j, k] is 1 if decoding the k-th element in the input sequence requires first decoding the j-th element and then the i-th element in the output sequence.

```python
In [1]: import torch

In [2]: import torch as th

In [3]: N = 3

In [4]: B = 1

In [5]: mask = torch.randint(0, 2, (N, N))

In [6]: mask.shape
Out[6]: torch.Size([3, 3])

In [7]: E_idx = torch.randint(0, 2, (1, N, N))

In [8]:     # (B, N)
...:     randn = th.rand(mask.shape).add_(5).abs()
...:     # Our mask=1 represents available data, vs the protein MPP's mask=1 which represents unavailable data
...:     decoding_order = th.argsort(-mask * randn)
...:     permutation_matrix_reverse = th.nn.functional.one_hot(
...:         decoding_order, num_classes=N
...:     ).float()
...:     # Calc mask from q to p, given the known q
...:     order_mask_backward = th.einsum(
...:         "ij, biq, bjp->bqp",
...:         th.tril(th.ones(N, N, dtype=permutation_matrix_reverse.dtype)),
...:         permutation_matrix_reverse,
...:         permutation_matrix_reverse,
...:     )
...:     mask_attend2 = th.gather(order_mask_backward, 2, E_idx)
...:     mask_1D = mask[..., None]
...:     mask_bw = mask_1D * mask_attend2
...:     mask_fw = mask_1D * (1 - mask_attend2)

In [9]: mask
Out[9]: 
tensor([[1, 1, 0],
    [1, 1, 0],
    [0, 1, 1]])

In [10]: E_idx
Out[10]: 
tensor([[[0, 1, 1],
     [1, 1, 0],
     [1, 1, 0]]])

In [11]: E_idx.shape
Out[11]: torch.Size([1, 3, 3])

In [12]: randn
Out[12]: 
tensor([[5.3629, 5.9851, 5.5263],
    [5.2165, 5.0775, 5.6979],
    [5.4468, 5.3644, 5.1452]])

In [13]: randn.shape
Out[13]: torch.Size([3, 3])

In [14]: decoding_order
Out[14]: 
tensor([[1, 0, 2],
    [0, 1, 2],
    [1, 2, 0]])

In [15]: permutation_matrix_reverse
Out[15]: 
tensor([[[0., 1., 0.],
     [1., 0., 0.],
     [0., 0., 1.]],

    [[1., 0., 0.],
     [0., 1., 0.],
     [0., 0., 1.]],

    [[0., 1., 0.],
     [0., 0., 1.],
     [1., 0., 0.]]])

In [16]: permutation_matrix_reverse.shape
Out[16]: torch.Size([3, 3, 3])

In [17]: order_mask_backward
Out[17]: 
tensor([[[1., 1., 0.],
     [0., 1., 0.],
     [1., 1., 1.]],

    [[1., 0., 0.],
     [1., 1., 0.],
     [1., 1., 1.]],

    [[1., 1., 1.],
     [0., 1., 0.],
     [0., 1., 1.]]])

In [18]: order_mask_backward.shape
Out[18]: torch.Size([3, 3, 3])

In [19]: mask_attend2
Out[19]: 
tensor([[[1., 1., 1.],
     [1., 1., 0.],
     [1., 1., 1.]]])

In [20]: mask_attend2.shape
Out[20]: torch.Size([1, 3, 3])

In [21]: mask_1D
Out[21]: 
tensor([[[1],
     [1],
     [0]],

    [[1],
     [1],
     [0]],

    [[0],
     [1],
     [1]]])

In [22]: mask_1D.shape
Out[22]: torch.Size([3, 3, 1])

In [23]: mask_bw
Out[23]: 
tensor([[[1., 1., 1.],
     [1., 1., 0.],
     [0., 0., 0.]],

    [[1., 1., 1.],
     [1., 1., 0.],
     [0., 0., 0.]],

    [[0., 0., 0.],
     [1., 1., 0.],
     [1., 1., 1.]]])

In [24]: mask_bw.shape
Out[24]: torch.Size([3, 3, 3])

In [25]: mask_fw
Out[25]: 
tensor([[[0., 0., 0.],
     [0., 0., 1.],
     [0., 0., 0.]],

    [[0., 0., 0.],
     [0., 0., 1.],
     [0., 0., 0.]],

    [[0., 0., 0.],
     [0., 0., 1.],
     [0., 0., 0.]]])

In [26]: mask_fw.shape
Out[26]: torch.Size([3, 3, 3])

In [27]:
```

It does not make sense to multiply `permutation_matrix_reverse` with itself since the one-hot operation does not produce negative values.

How is `mask_attend2` different from what `sqmask = mask[..., None, :] * mask[..., None]` is trying to achieve as [adjacency matrix](https://github.com/A4Bio/PiFold/blob/f111d66ed4c184b9614ed670da26469dcc35e588/methods/prodesign_model.py#L109) ?

**Comment by simonlevine** [comment]
+1. These masks aren't even used in later operations.

**Comment by gaozhangyang** [comment]
These codes are segments from ProteinMPNN, utilized to explore the impact of the autoregressive scheme on the model's performance. However, they are not employed in the final version of PiFold.
